pkg/snapstore/abs_snapstore.go

```diff
@@ -56,32 +59,63 @@ type absCredentials struct {
 	StorageAccount string `json:"storageAccount"`
 }
 
-// NewABSSnapStore create new ABSSnapStore from shared configuration with specified bucket
+// NewABSSnapStore creates a new ABSSnapStore using a shared configuration and a specified bucket
 func NewABSSnapStore(config *brtypes.SnapstoreConfig) (*ABSSnapStore, error) {
 	storageAccount, storageKey, err := getCredentials(getEnvPrefixString(config.IsSource))
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to get credentials: %v", err)
 	}
 
 	credentials, err := azblob.NewSharedKeyCredential(storageAccount, storageKey)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create shared key credentials: %v", err)
 	}
 
-	p := azblob.NewPipeline(credentials, azblob.PipelineOptions{
+	pipeline := azblob.NewPipeline(credentials, azblob.PipelineOptions{
 		Retry: azblob.RetryOptions{
 			TryTimeout: downloadTimeout,
 		}})
-	u, err := url.Parse(fmt.Sprintf("https://%s.%s", storageAccount, brtypes.AzureBlobStorageHostName))
+
+	blobURL, err := ConstructBlobServiceURL(credentials)
 	if err != nil {
-		return nil, fmt.Errorf("failed to parse service url: %v", err)
+		return nil, err
 	}
-	serviceURL := azblob.NewServiceURL(*u, p)
+
+	serviceURL := azblob.NewServiceURL(*blobURL, pipeline)
 	containerURL := serviceURL.NewContainerURL(config.Container)
 
 	return GetABSSnapstoreFromClient(config.Container, config.Prefix, config.TempDir, config.MaxParallelChunkUploads, config.MinChunkSize, &containerURL)
 }
 
+// ConstructBlobServiceURL constructs the Blob Service URL based on the activation status of the Azurite Emulator.
+// It checks the environment variables for emulator configuration and constructs the URL accordingly.
+// The function expects two environment variables:
+// - EMULATOR_ENABLED: Indicates whether the Azurite Emulator is enabled (expects "true" or "false").
+// - AZURE_STORAGE_API_ENDPOINT: Specifies the Azurite Emulator endpoint when the emulator is enabled.
+func ConstructBlobServiceURL(credentials *azblob.SharedKeyCredential) (*url.URL, error) {
+	defaultURL, err := url.Parse(fmt.Sprintf("https://%s.%s", credentials.AccountName(), brtypes.AzureBlobStorageHostName))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse default service URL: %w", err)
+	}
+	emulatorEnabled, ok := os.LookupEnv(EnvEmulatorEnabled)
+	if !ok {
+		return defaultURL, nil
+	}
+	isEmulator, err := strconv.ParseBool(emulatorEnabled)
+	if err != nil {
+		return nil, fmt.Errorf("invalid value for %s: %s, error: %w", EnvEmulatorEnabled, emulatorEnabled, err)
+	}
+	if !isEmulator {
+		return defaultURL, nil
+	}
+	endpoint, ok := os.LookupEnv(AzuriteEndpoint)
+	if !ok {
+		return nil, fmt.Errorf("%s environment variable not set while %s is true", AzuriteEndpoint, EnvEmulatorEnabled)
+	}
+	// Application protocol (http or https) is determined by the user of the Azurite, not by this function.
+	return url.Parse(fmt.Sprintf("%s/%s", endpoint, credentials.AccountName()))
```

Comment:
Can we re-write this function like this, it becomes easy to understand.
```go
	if emulatorEnabled, ok := os.LookupEnv(EnvEmulatorEnabled); ok {
		isEmulator, err := strconv.ParseBool(emulatorEnabled)
		if err != nil {
			return nil, fmt.Errorf("invalid value for %s: %s, error: %w", EnvEmulatorEnabled, emulatorEnabled, err)
		}
		if isEmulator {
			endpoint, ok := os.LookupEnv(AzuriteEndpoint)
			if !ok {
				return nil, fmt.Errorf("%s environment variable not set while %s is true", AzuriteEndpoint, EnvEmulatorEnabled)
			}
			// Application protocol (http or https) is determined by the user of the Azurite, not by this function.
			return url.Parse(fmt.Sprintf("%s/%s", endpoint, credentials.AccountName()))
		}

	}

	defaultURL, err := url.Parse(fmt.Sprintf("https://%s.%s", credentials.AccountName(), brtypes.AzureBlobStorageHostName))
	if err != nil {
		return nil, fmt.Errorf("failed to parse default service URL: %w", err)
	}
	return defaultURL, nil

```
But, no strong opinion I will this decision to you.

Reply:
Thanks @ishan16696 for your suggestion! The flow initially was identical to the suggestion - however @seshachalam-yv suggested writing the flow in a non-nested fashion as it is currently - see this [review comment](https://github.com/gardener/etcd-backup-restore/pull/699?notification_referrer_id=NT_kwDOAw8xCrM4OTQxNTkwNjMwOjUxMzI3MjQy#discussion_r1480925343). The idea was to have an early return whenever an error is come across. I prefer if we keep the code as is.

Reply:
> The idea was to have an early return whenever an error is come across. I prefer if we keep the code as is.

What make you think so ? Please have a look at this: https://github.com/gardener/etcd-backup-restore/pull/699#discussion_r1487381753, did you notice any place in which we are getting an error still continuing and not returning an error 🤔

Reply:
Sorry for not explaining what I intended to say more clearly @ishan16696!

The https://github.com/gardener/etcd-backup-restore/pull/699#discussion_r1487381753 does not have any place where we continue after an error occurs. It is correct code.

What I wanted to say was to not have a nested `if-else` and have a flatter structure for the code. "Early return" happens in both cases - the current code and your suggestion as well. The previous suggestion by @seshachalam-yv was to not have a nested `if-else`, that is all.

Went ahead and merged the code yesterday since you had approved these changes - should have waited for your response, apologies!

Reply:
no worries.